README.md
# process-automation README

This is the README for your extension "process-automation". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

You can author your README using Visual Studio Code. Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux).
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux).
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets.

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "process-automation" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/paTree.ts


import * as vscode from 'vscode';

export class PrcAutTree {
    constructor(context:vscode.ExtensionContext){
        const vweTree =vscode.window.createTreeView('paTree',{treeDataProvider:aNodeWithIdTreeDataProvider(),showCollapseAll:true});
		context.subscriptions.push(vweTree);
		vscode.commands.registerCommand('paTree.reveal', async () => {
			const key = await vscode.window.showInputBox({ placeHolder: 'Type the label of the item to reveal' });
			if (key) {
				await vweTree.reveal({ key }, { focus: true, select: false, expand: true });
			}
		});
		vscode.commands.registerCommand('paTree.changeTitle', async () => {
			const title = await vscode.window.showInputBox({ prompt: 'Type the new title for the Test View', placeHolder: vweTree.title });
			if (title) {
				vweTree.title = title;
			}
		});
    }
}

const tree:any = {

};
const nodes:any={};

function aNodeWithIdTreeDataProvider() : vscode.TreeDataProvider<{ key: string }> {
    return {
		getChildren: (element: { key: string }): { key: string }[] => {
			return getChildren(element ? element.key : undefined).map(key => getNode(key));
		},
		getTreeItem: (element: { key: string }): vscode.TreeItem => {
			const treeItem = getTreeItem(element.key);
			treeItem.id = element.key;
			return treeItem;
		},
		getParent: ({ key }: { key: string }): { key: string } | undefined => {
			const parentKey = key.substring(0, key.length - 1);
			return parentKey ? new Key(parentKey) : undefined;
		}

    };
}

function getChildren(key: string | undefined): string[] {
	if (!key) {
		return Object.keys(tree);
	}
	const treeElement = getTreeElement(key);
	if (treeElement) {
		return Object.keys(treeElement);
	}
	return [];
}

function getTreeItem(key: string): vscode.TreeItem {
	const treeElement = getTreeElement(key);
	// An example of how to use codicons in a MarkdownString in a tree item tooltip.
	const tooltip = new vscode.MarkdownString(`$(zap) Tooltip for ${key}`, true);
	return {
		label: /**vscode.TreeItemLabel**/<any>{ label: key, highlights: key.length > 1 ? [[key.length - 2, key.length - 1]] : void 0 },
		tooltip,
		collapsibleState: treeElement && Object.keys(treeElement).length ? vscode.TreeItemCollapsibleState.Collapsed : vscode.TreeItemCollapsibleState.None
	};
}

function getTreeElement(element: string): any {
	let parent = tree;
	for (let i = 0; i < element.length; i++) {
		parent = parent[element.substring(0, i + 1)];
		if (!parent) {
			return null;
		}
	}
	return parent;
}

function getNode(key: string): { key: string } {
	if (!nodes[key]) {
		nodes[key] = new Key(key);
	}
	return nodes[key];
}

class Key {
	constructor(readonly key: string) { }
}

webview-ui/main.js
// Get access to the VS Code API from within the webview context
const vscode = acquireVsCodeApi();

// Just like a regular webpage we need to wait for the webview
// DOM to load before we can reference any of the HTML elements
// or toolkit components
window.addEventListener("load", main);

function main() {
    setVSCodeMessageListener();
    vscode.postMessage({ command: "requestNoteData" });

    const saveButton = document.getElementById("submit-button");
    saveButton.addEventListener("click", () => saveNote());
    console.log('save add listener');
    // svg
    var cnvViewer=d3.select("#paViewer");
    cnvViewer.append("circle")
        .style("stroke", "gray")
        .style("fill", "white")
        .attr("r", 20)
        .attr("cx", 50)
        .attr("cy", 50);
    cnvViewer.append("circle")
        .style("stroke", "gray")
        .style("fill", "white")
        .attr("r", 20)
        .attr("cx", 120)
        .attr("cy", 50);
    cnvViewer.append("circle")
        .style("stroke", "gray")
        .style("fill", "white")
        .attr("r", 25)
        .attr("cx", 200)
        .attr("cy", 50);
    var svgSub=cnvViewer.append("svg");
    var circle=svgSub.selectAll("circle")
          .data([5,7,9,5,3,10,20])
          //.data([10,10,10,10,5,5]);
          //.data([10,10,10,10,5,5],function(d){return d;});
          .enter()
          .append("circle")
          .style("stroke", "black")
          //.style("fill","url(#red-light-off) #002804")
          //.style("fill","url(#yellow-light-off) #002804")
          .style("fill","url(#green-light-off) #002804")
          //.style("fill","white")
          .attr("cy",100)
          .attr("cx",function(d,i){return (i*40)+50;})
          .attr("r",function(d){return 20+d;});
    circle.exit().remove();
    console.log('svg');


}

// Stores the currently opened note info so we know the ID when we update it on save
let openedNote;

function setVSCodeMessageListener() {
  window.addEventListener("message", (event) => {
    const command = event.data.command;
    const noteData = JSON.parse(event.data.payload);

    switch (command) {
      case "receiveDataInWebview":
        openedNote = noteData;
        renderTags(openedNote.tags);
        break;
    }
  });
}

function saveNote() {
  const titleInputValue = document.getElementById("title").value;
  const noteInputValue = document.getElementById("content").value;
  const tagsInputValue = document.getElementById("tags-input").value;

  let tagsInputList = [];
  if (tagsInputValue.length > 0) {
    tagsInputList = tagsInputValue.split(",").map((tag) => tag.trim());
  }

  const noteToUpdate = {
    id: openedNote.id,
    title: titleInputValue,
    content: noteInputValue,
    tags: tagsInputList,
  };

  vscode.postMessage({ command: "updateNote", note: noteToUpdate });
}

function renderTags(tags) {
  const tagsContainer = document.getElementById("tags-container");
  clearTagGroup(tagsContainer);
  if (tags.length > 0) {
    addTagsToTagGroup(tags, tagsContainer);
    tagsContainer.style.marginBottom = "2rem";
  } else {
    // Remove tag container bottom margin if there are no tags
    tagsContainer.style.marginBottom = "0";
  }
}

function clearTagGroup(tagsContainer) {
  while (tagsContainer.firstChild) {
    tagsContainer.removeChild(tagsContainer.lastChild);
  }
}

function addTagsToTagGroup(tags, tagsContainer) {
  for (const tagString of tags) {
    const vscodeTag = document.createElement("vscode-tag");
    vscodeTag.textContent = tagString;
    tagsContainer.appendChild(vscodeTag);
  }
}


src/ui/getWebviewContent.ts
import { Webview, Uri } from "vscode";
import { Note } from "../types/Note";
import { getUri } from "../utilities/getUri";

/**
 * Defines and returns the HTML that should be rendered within the notepad webview panel.
 *
 * @remarks This is also the place where references to CSS and JavaScript files/packages
 * (such as the Webview UI Toolkit) are created and inserted into the webview HTML.
 *
 * @param webview A reference to the extension webview
 * @param extensionUri The URI of the directory containing the extension
 * @param note An object representing a notepad note
 * @returns A template string literal containing the HTML that should be
 * rendered within the webview panel
 */
export function getWebviewContent(webview: Webview, extensionUri: Uri, note: Note) {
  const uriToolkit = getUri(webview, extensionUri, ["node_modules",
                          "@vscode","webview-ui-toolkit","dist","toolkit.js",
                          ]);
  const uriD3 = getUri(webview, extensionUri, ["node_modules",
                          "d3","dist","d3.min.js",
                          ]);
  const uriStyle = getUri(webview, extensionUri, ["webview-ui", "style.css"]);
  const uriMain = getUri(webview, extensionUri, ["webview-ui", "main.js"]);

  const formattedTags = note.tags ? note.tags.join(", ") : null;

  webview.onDidReceiveMessage((message) => {
    const command = message.command;
    switch (command) {
      case "requestNoteData":
        webview.postMessage({
          command: "receiveDataInWebview",
          payload: JSON.stringify(note),
        });
        break;
    }
  });

  return /*html*/ `
    <!DOCTYPE html>
    <html lang="en">
      <head>
          <meta charset="UTF-8">
          <meta name="viewport" content="width=device-width, initial-scale=1.0">
          <script type="module" src="${uriToolkit}"></script>
          <script type="module" src="${uriMain}"></script>
          <script type="module" src="${uriD3}"></script>
          <link rel="stylesheet" href="${uriStyle}">
          <title>${note.title}</title>
      </head>
      <body id="webview-body">
        <svg id="paViewer" class="viewer" 
          viewBox="20 20 400 320" preserveAspectRatio="xMinYMin meet"width="800px" height="200px">
          <defs>
            <circle id="light" cx="70" r="30" />
            <radialGradient id="red-light-on" fx="0.45" fy="0.4">
            <stop stop-color="orange" offset="0.1"/>
            <stop stop-color="red" offset="0.8"/>
            <stop stop-color="brown" offset="1.0"/>
         </radialGradient>
         <radialGradient id="red-light-off" fx="0.45" fy="0.4">
             <stop stop-color="maroon" offset="0"/>
             <stop stop-color="#220000" offset="0.7"/>
             <stop stop-color="black" offset="1.0"/>
         </radialGradient>
         <radialGradient id="yellow-light-on" fx="0.45" fy="0.4">
             <stop stop-color="yellow" offset="0.1"/>
             <stop stop-color="orange" offset="0.6"/>
             <stop stop-color="darkOrange" offset="1.0"/>
         </radialGradient>
         <radialGradient id="yellow-light-off" fx="0.45" fy="0.4">
             <stop stop-color="#A06000" offset="0"/>
             <stop stop-color="#804000" offset="0.7"/>
             <stop stop-color="#502000" offset="1"/>
         </radialGradient>
         <radialGradient id="green-light-on" fx="0.45" fy="0.4">
             <stop stop-color="#88FF00" offset="0.1"/>
             <stop stop-color="forestGreen" offset="0.7"/>
             <stop stop-color="darkGreen" offset="1.0"/>
         </radialGradient>
         <radialGradient id="green-light-off" fx="0.45" fy="0.4">
             <stop stop-color="#007000" offset="0"/>
             <stop stop-color="#004000" offset="0.6"/>
             <stop stop-color="#001000" offset="1.0"/>
         </radialGradient>
         <linearGradient id="metal" spreadMethod="repeat"
                         gradientTransform="scale(0.7) rotate(75)">
             <stop stop-color="#808080" offset="0"/>
             <stop stop-color="#404040" offset="0.25"/>
             <stop stop-color="#C0C0C0" offset="0.35"/>
             <stop stop-color="#808080" offset="0.5"/>
             <stop stop-color="#E0E0E0" offset="0.7"/>
             <stop stop-color="#606060" offset="0.75"/>
             <stop stop-color="#A0A0A0" offset="0.9"/>
             <stop stop-color="#808080" offset="1"/>
         </linearGradient>
        </defs>
        <rect x="20" y="20" width="100" height="280" 
              fill="url(gradients.svg#metal) silver" 
              stroke="black" stroke-width="3" /> 
          <g stroke="black" stroke-width="2">
              <g class="red light" >
                  <use xlink:href="#light" y="80" 
                       fill="url(#red-light-off) maroon" />
                  <use class="lit" xlink:href="#light" y="80" 
                       fill="url(#red-light-on) red" >
                      <title>STOP</title>
                  </use>
              </g>
              <g class="yellow light">
                  <use xlink:href="#light" y="160" 
                       fill="url(#yellow-light-off) #705008" />
                  <use class="lit" xlink:href="#light" y="160" 
                       fill="url(#yellow-light-on) yellow" />
              </g>
              <g class="green light">
                  <use xlink:href="#light" y="240" 
                       fill="url(#green-light-off) #002804" />
                  <use class="lit" xlink:href="#light" y="240" 
                       fill="url(#green-light-on) lime" >
                      <title>GO</title>
                  </use>
              </g>
          </g>
        </svg>
        <header>
          <h1>${note.title}</h1>
          <div id="tags-container"></div>
        </header>
        <section id="notes-form">
          <vscode-text-field id="title" value="${note.title}" placeholder="Enter a name">Title</vscode-text-field>
          <vscode-text-area id="content"value="${note.content}" placeholder="Write your heart out, Shakespeare!" resize="vertical" rows=15>Note</vscode-text-area>
          <vscode-text-field id="tags-input" value="${formattedTags}" placeholder="Add tags separated by commas">Tags</vscode-text-field>
          <vscode-button id="submit-button">Save</vscode-button>
        </section>
      </body>
    </html>
  `;
}


src/providers/PrcAutDataProvider.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import { AnyRecord } from 'dns';
import { lch } from 'd3';

interface IfcPrcAutJson {
    prcAutJson:{ [k: string]: any };
}

export class PrcAutDataProvider implements vscode.TreeDataProvider<PrcAutTreeItem> {

	private _onDidChangeTreeData: vscode.EventEmitter<PrcAutTreeItem | undefined | void> = new vscode.EventEmitter<PrcAutTreeItem | undefined | void>();
	readonly onDidChangeTreeData: vscode.Event<PrcAutTreeItem | undefined | void> = this._onDidChangeTreeData.event;

	constructor(private workspaceRoot: string | undefined) {
	}

	refresh(): void {
		this._onDidChangeTreeData.fire();
	}

	getTreeItem(element: PrcAutTreeItem): vscode.TreeItem {
		return element;
	}
    /**
     * read process automation json PrcAut.pa.json
     */
	getChildren(element?: PrcAutTreeItem): Thenable<PrcAutTreeItem[]> {
            if (!this.workspaceRoot) {
			vscode.window.showInformationMessage('No PrcAutTreeItem in empty workspace');
			return Promise.resolve([]);
		}

		if (element) {
            return Promise.resolve(this.getTreeItemByPrcAutJson(element.getPrcAutJson()));
		} else {
			const packageJsonPath = path.join(this.workspaceRoot, 'PrcAut.pa.json');
			if (this.pathExists(packageJsonPath)) {
				return Promise.resolve(this.getPrcAutInJson(packageJsonPath));
			} else {
				vscode.window.showInformationMessage('Workspace has no PrcAut.pa.json');
				return Promise.resolve([]);
			}
		}
	}
	/**
	 * Given the path to process automation json,
     * read content and build tree items
	 */
	private getPrcAutInJson(jsonPath: string): PrcAutTreeItem[] {
		const workspaceRoot = this.workspaceRoot;
		if (this.pathExists(jsonPath) && workspaceRoot) {
			const prcAutJson = JSON.parse(fs.readFileSync(jsonPath, 'utf-8'));
            
            return this.getTreeItemByPrcAutJson(prcAutJson);
		} else {
			return [];
		}
	}
    /**
     * evaluate json dictionary, looks for valid type set
     * checks if element has a child or children to create tree item
     * @param prcAutJson 
     * @returns list of tree items
     */
    private getTreeItemByPrcAutJson(prcAutJson:{ [k: string]: any }):PrcAutTreeItem[]{
        if (prcAutJson){

            let lChild:PrcAutTreeItem[]=[];
            for (const [sKey,o] of Object.entries(prcAutJson)){
                console.log(sKey);
                console.log(prcAutJson[sKey]);
                let sType=o._type;
                if (sType){
                    let iHasChild:number=0;
                    for (const [sKey,oC] of Object.entries(o)){
                        if (!sKey.startsWith('_')){
                            iHasChild++;
                        }
                    }
                    if (iHasChild>0){
                        lChild.push(new PrcAutTreeItem(sKey, sType,o, vscode.TreeItemCollapsibleState.Collapsed));
                    }else{
                        lChild.push(new PrcAutTreeItem(sKey, sType,o, vscode.TreeItemCollapsibleState.None));
                    }
                }
            }
            return lChild;
        }else{
            return [];
        }
    }
    /**
     * check if file exists
     * @param p 
     * @returns 
     */
	private pathExists(p: string): boolean {
		try {
			fs.accessSync(p);
		} catch (err) {
			return false;
		}

		return true;
	}
}

export class PrcAutTreeItem extends vscode.TreeItem {
    prcAutJson?:{ [k: string]: any };
	constructor(
		public readonly label: string,
		private readonly type: string,
		private readonly prcAutJson:{ [k: string]: any },
		public readonly collapsibleState: vscode.TreeItemCollapsibleState,
		public readonly command?: vscode.Command
	) {
		super(label, collapsibleState);
        this.prcAutJson=prcAutJson;
        this.version='';
		this.tooltip = `${this.label}-${this.version}`;
		this.description = this.version;
	}
    public getPrcAutJson():{ [k: string]: any } {
        return this.prcAutJson;
    }
	iconPath = {
		light: path.join(__filename, '..', '..', 'resources', 'light', 'dependency.svg'),
		dark: path.join(__filename, '..', '..', 'resources', 'dark', 'dependency.svg')
	};

	contextValue = 'dependency';
}


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';

import { TestView } from './testView';

// https://github.com/microsoft/vscode-webview-ui-toolkit-samples/tree/main/default/notepad
// extensions.ts
import { v4 as uuidv4 } from "uuid";
import { NotepadDataProvider } from "./providers/NotepadDataProvider";
import { getWebviewContent } from "./ui/getWebviewContent";
import { Note } from "./types/Note";

import * as d3 from "d3";
//import * as _d3 from "d3";
import { select, Selection} from 'd3-selection';
import { getPAviewerContent } from './paViewer';

//import {walk} from './utilities/walk';
import { PrcAutDataProvider } from "./providers/PrcAutDataProvider";

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
	// +++++ beg: sample notepad
	let notes: Note[] = [];
	let panel: vscode.WebviewPanel | undefined = undefined;
  
	const notepadDataProvider = new NotepadDataProvider(notes);
  
	const treeView = vscode.window.createTreeView("notepad.notesList", {
	  treeDataProvider: notepadDataProvider,
	  showCollapseAll: false,
	});
	const openNote = vscode.commands.registerCommand("notepad.showNoteDetailView", () => {
		const selectedTreeViewItem = treeView.selection[0];
		const matchingNote = notes.find((note) => note.id === selectedTreeViewItem.id);
		if (!matchingNote) {
			vscode.window.showErrorMessage("No matching note found");
		  return;
		}
	
		// If no panel is open, create a new one and update the HTML
		if (!panel) {
		  panel = vscode.window.createWebviewPanel("noteDetailView", matchingNote.title, vscode.ViewColumn.One, {
			enableScripts: true,
		  });
		}
	
		// If a panel is open, update the HTML with the selected item's content
		panel.title = matchingNote.title;
		panel.webview.html = getWebviewContent(panel.webview, context.extensionUri, matchingNote);
	
		// If a panel is open and receives an update message, update the notes array and the panel title/html
		panel.webview.onDidReceiveMessage((message) => {
		  const command = message.command;
		  const note = message.note;
		  switch (command) {
			case "updateNote":
			  const updatedNoteId = note.id;
			  const copyOfNotesArray = [...notes];
			  const matchingNoteIndex = copyOfNotesArray.findIndex((note) => note.id === updatedNoteId);
			  copyOfNotesArray[matchingNoteIndex] = note;
			  notes = copyOfNotesArray;
			  notepadDataProvider.refresh(notes);
			  panel
				? ((panel.title = note.title),
				  (panel.webview.html = getWebviewContent(panel.webview, context.extensionUri, note)))
				: null;
			  break;
		  }
		});
	
		panel.onDidDispose(
		  () => {
			// When the panel is closed, cancel any future updates to the webview content
			panel = undefined;
		  },
		  null,
		  context.subscriptions
		);
	  });
	
	  const createNote = vscode.commands.registerCommand("notepad.createNote", () => {
		const id = uuidv4();
	
		const newNote: Note = {
		  id: id,
		  title: "New note",
		  content: "",
		  tags: ["Personal"],
		};
	
		notes.push(newNote);
		notepadDataProvider.refresh(notes);
	  });
	
	  const deleteNote = vscode.commands.registerCommand("notepad.deleteNote", (node: Note) => {
		const selectedTreeViewItem = node;
		const selectedNoteIndex = notes.findIndex((note) => note.id === selectedTreeViewItem.id);
		notes.splice(selectedNoteIndex, 1);
		notepadDataProvider.refresh(notes);
	
		// Close the panel if it's open
		panel?.dispose();
	  });
	
	// Add commands to the extension context
	context.subscriptions.push(openNote);
	context.subscriptions.push(createNote);
	context.subscriptions.push(deleteNote);
	// ----- end: sample notepad
	// +++++ beg:PrcAut
	const rootPath = (vscode.workspace.workspaceFolders && (vscode.workspace.workspaceFolders.length > 0))
		? vscode.workspace.workspaceFolders[0].uri.fsPath : undefined;
	const pvrPrcAut=new PrcAutDataProvider(rootPath);
	vscode.window.registerTreeDataProvider('PrcAut', pvrPrcAut);
	vscode.commands.registerCommand('PrcAut.refresh', () => pvrPrcAut.refresh());
	// ----- end:PrcAut
	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "process-automation" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('process-automation.helloWorld', () => {
		// The code you place here will be executed every time your command is executed
		// Display a message box to the user
		vscode.window.showInformationMessage('Hello World from process-automation!');
	});

	// WebView
	//let disposableWebView = vscode.commands.registerCommand('process-automation.Viewer',() => {
	//	const pnPAviewer=vscode.window.createWebviewPanel(
	//		'pa.Viewer',
	//		'Process Automation Viewer',
	//		vscode.ViewColumn.One,
	//		{}
	//	);
	//	pnPAviewer.webview.html=getPAviewerContent();
		//var cnvViewer=d3.select("#paViewer");
		/*var circle=d3.selectAll("circle")
        	.data([10,5,5])
    	    .enter()
	        .append("circle");
		circle.attr("cy",40);
		circle.attr("cx",function(d,i){return (i*50)+20;});
		circle.attr("r",function(d){return d;});*/
	//});
	// Test View
	//new TestView(context);

	context.subscriptions.push(disposable);
	//context.subscriptions.push(disposableWebView);
}

// This method is called when your extension is deactivated
export function deactivate() {}
